apps/aehttp/src/aehttp_dispatch_rosetta.erl

```diff
@@ -0,0 +1,659 @@
+            _ ->
+                throw(invalid_pubkey)
+        end
+    catch
+        invalid_network ->
+            {200, [], rosetta_error_response(?ROSETTA_ERR_INVALID_NETWORK)};
+        block_not_found ->
+            {200, [], rosetta_error_response(?ROSETTA_ERR_BLOCK_NOT_FOUND)};
+        chain_too_short ->
+            {200, [], rosetta_error_response(?ROSETTA_ERR_CHAIN_TOO_SHORT)};
+        invalid_pubkey ->
+            {200, [], rosetta_error_response(?ROSETTA_ERR_TX_INVALID_ACCOUNT)}
+    end;
+handle_request_(accountCoins, _, _Context) ->
+    {501, [], #{}};
+%% Data API - Block
+handle_request_(block,
+                #{'BlockRequest' :=
+                      #{<<"network_identifier">> :=
+                            #{<<"blockchain">> := <<"aeternity">>, <<"network">> := Network}} =
+                          Req},
+                _Context) ->
+    try
+        case aec_governance:get_network_id() of
+            Network ->
+                ok;
+            _ ->
+                throw(invalid_network)
+        end,
+        Block = retrieve_block_from_partial_block_identifier(Req),
+        BlockFmt = format_block(Block),
+        Resp = #{<<"block">> => BlockFmt},
+        {200, [], Resp}
+    catch
+        invalid_network ->
+            {200, [], rosetta_error_response(?ROSETTA_ERR_INVALID_NETWORK)};
+        block_not_found ->
+            {200, [], rosetta_error_response(?ROSETTA_ERR_BLOCK_NOT_FOUND)};
+        chain_too_short ->
+            {200, [], rosetta_error_response(?ROSETTA_ERR_CHAIN_TOO_SHORT)};
+        {dry_run_err, Err} ->
+            dry_run_err(Err)
+    end;
+handle_request_(blockTransaction,
+                #{'BlockTransactionRequest' :=
+                      #{<<"block_identifier">> := BlockIdentifier,
+                        <<"network_identifier">> := #{<<"blockchain">> := <<"aeternity">>},
+                        <<"transaction_identifier">> := #{<<"hash">> := TxHash}}},
+                _Context) ->
+    try
+        %%
+        %% Fetch block
+        %%
+        Block = retrieve_block(BlockIdentifier),
+        %%
+        %% Decode the transaction hash
+        %%
+        TxHashInternal =
+            case aeapi:decode(TxHash) of
+                {tx_hash, TxHash0} ->
+                    TxHash0;
+                _ ->
+                    throw(tx_not_found)
+            end,
+        %%
+        %% Fetch the transaction, ensuring it was in the specified block
+        %%
+        {SignedTx, Microblock} =
+            case aec_db:find_tx_with_location(TxHashInternal) of
+                none ->
+                    throw(block_not_found);
+                {mempool, _} ->
+                    throw(block_not_found);
+                {BlockHash, STx} ->
+                    %% Tx was in this microblock. Check this microblock is in the
+                    %% Requested keyblock
+                    case aec_chain:get_block(BlockHash) of
+                        {ok, MBlock} ->
```

Comment:
Should be done within one activity.

Reply:
As discussed will leave this as is for the moment. It's not performance critical and is clear enough